File: examples/modular/order-management/python/_cott_impl/store/order/calculate_order.py
from cott_runtime import Err, Ok, Result
from store.catalog import find_item
from store.catalog_types import Catalog
from store.order import validate_line
from store.order_types import Order, OrderError, OrderError_EmptyOrder, OrderError_ItemUnavailable, OrderReceipt


def calculate_order(catalog: Catalog, order: Order) -> Result[OrderReceipt, OrderError]:
    if len(order.lines) == 0:
        return Err(error=OrderError_EmptyOrder())

    total_items = 0
    total_cents = 0
    for line in order.lines:
        validated = validate_line(line)
        if isinstance(validated, Err):
            return Err(error=validated.error)

        found = find_item(catalog, validated.value.sku)
        if isinstance(found, Err):
            return Err(error=OrderError_ItemUnavailable(cause=found.error))

        total_items += validated.value.quantity
        total_cents += found.value.price_cents * validated.value.quantity

    return Ok(value=OrderReceipt(order_id=order.order_id, total_items=total_items, total_cents=total_cents))

File: examples/kotlin/modular/order-management/kotlin/cott_impl/store/order/calculate_order.kt
package cott_impl.store.order

internal fun calculate_order(catalog: store.catalog.Catalog, order: store.order.Order): cott_runtime.CottResult<store.order.OrderReceipt, store.order.OrderError> {
    if (order.lines.isEmpty()) {
        return cott_runtime.Err(store.order.OrderError.EmptyOrder)
    }

    var totalItems = 0u
    var totalCents = 0uL
    for (line in order.lines) {
        val valid = when (val validation = store.order.validate_line(line)) {
            is cott_runtime.Err -> return cott_runtime.Err(validation.error)
            is cott_runtime.Ok -> validation.value
        }
        val item = when (val lookup = store.catalog.find_item(catalog, valid.sku)) {
            is cott_runtime.Err -> return cott_runtime.Err(store.order.OrderError.ItemUnavailable(lookup.error))
            is cott_runtime.Ok -> lookup.value
        }
        totalItems += valid.quantity
        totalCents += item.price_cents * valid.quantity.toULong()
    }

    return cott_runtime.Ok(store.order.OrderReceipt(order.order_id, totalItems, totalCents))
}
